### ISSUE TITLE
DAX Formatting Not Quite Right in 3.1.6

### ISSUE THREAD

**Issue by edhans** [TabularEditor/TabularEditor3#306]
This is how a measure is formatted using DAXFormatter.com using Short LInes
<img width="563" alt="image" src="https://user-images.githubusercontent.com/45298358/143152052-28e3311f-8acf-47ee-bb85-9c2ff92036a3.png">

This is how it is formatted using Short LInes using the new internal formatter.
<img width="312" alt="image" src="https://user-images.githubusercontent.com/45298358/143152159-36b2ad05-452d-4790-997a-95da71d21b6f.png">

The CROSSFILTER might be ok, though I prefer DAXFormatter keeping that kind of thing on one line, but the MAX() function should not take up 3 lines. Should be on one line.

There is also a bug after a /* comment */ line.
<img width="495" alt="image" src="https://user-images.githubusercontent.com/45298358/143152322-cbe8990c-ea58-448a-b628-0b47331fab27.png">

I kept pressing the formatter and it kept adding new rows and never removing them. The only fix is to manually remove them, or switch back to DAXFormatter.com which removed all of it.

**Comment by edhans** [comment]
There is also something not right with using this with a macro. I use the following macro to format my code:
```
/*  Cycle over all measures in model format using
    DAX Formatter with Short Lines, then add 1 line
    feed to the start of the measure
    Then copies the DAX to the description so it
    is available in the hover over */
FormatDax ( Model.AllMeasures ,  true);
foreach  (var  m  in  Model.AllMeasures )
     {
         m.Expression   =   '\n'   +   m.Expression ;
         m.Description   =   m.Expression ;
     }

```
It inserts a blank , then the expression. This is so I don't get the first part of the expression in the = line in Power BI.
In other words, I want:
Some Measure =
   CALCULATE(
       something.....
       
not:
Some Measure = CALCULATE(
    something....

but now the code generates what is below - without the extra line, and it isn't fixing the string in the description field either. In tests I ran the macro 7-8 times, and it just kept adding lines to the code and the format does not match the expression.    

Switching everything back to DAXFormatter.com fixes all of it.

<img width="1122" alt="image" src="https://user-images.githubusercontent.com/45298358/143152634-69500c6e-add4-48ca-936e-4cc27f7d6375.png">

**Comment by otykier** [comment]
We will look into the issues highlighted in your 1st post.

Regarding your 2nd post, this is related to #307, since the implicit call to `CallDaxFormatter()` (which happens after the script finishes execution) doesn't work with the built-in formatter. Instead, consider changing your script to the following:

```csharp
/*  Cycle over all measures in model format using
    DAX Formatter with Short Lines, then add 1 line
    feed to the start of the measure
    Then copies the DAX to the description so it
    is available in the hover over */
foreach  (var  m  in  Model.AllMeasures )
     {
         m.Expression   =   '\n'   + FormatDax(m.Expression, shortFormat: true);
         m.Description   =   m.Expression ;
     }
```

**Comment by PhZotz** [comment]
Hi Daniel, thanks for providing this workaround, I have started to rely so much on a script to format all measures at once.
I was wondering if its possible to apply the same to all calculated columns? I couldn't get this running with my very limited scripting skills.
Ideally I would like to have one script that formats all my measures and calculated columns.

Thanks!

**Comment by edhans** [comment]
You would have to turn on unsupported operations Phzotz since MS doesn't support the modification of calculated columns in a PBIX file by external tools yet, which would include formatting the DAX code.

**Comment by PhZotz** [comment]
Hi edhans, yes I have that enabled and it works fine on single colums. It would be nice to have it in a macro to apply it to all columns at once.
I have just realized that DAX scripting now supports columns as well so I can just open the DAX script of the entire model press F6 and I'm done.
Still curious if its possible in a C# script, too.

**Comment by samaguire** [comment]
> Hi edhans, yes I have that enabled and it works fine on single colums. It would be nice to have it in a macro to apply it to all columns at once. I have just realized that DAX scripting now supports columns as well so I can just open the DAX script of the entire model press F6 and I'm done. Still curious if its possible in a C# script, too.

Yes, I have done this. Formats measures, columns, tables and calc group items:
 https://gist.github.com/samaguire/e8f60ea57e967070a5325559ae245f9d

It needs to be adapted to work with the internal formatter, or you can just change preferences to keep using daxformatter.com like I've done.

**Comment by PhZotz** [comment]
That's exactly what I was looking for, thank you samaguire 👏

**Comment by samaguire** [comment]
> That's exactly what I was looking for, thank you samaguire 👏

I'll look at converting the code to get it working with the internal formatter if that is the option you've chosen. Stay tuned for an update.

**Comment by samaguire** [comment]
New code, but it seems to be ignoring the preferences for the internal formatter.

``` DAX
/*

Formatting of calculated tables and calculated columns is unsupported in Power BI Desktop as at 25 Aug 2021
and will be ignored if unsupported features are not enabled in Tabular Editor

NB: This will always use the internal formatter (https://github.com/TabularEditor/TabularEditor3/issues/307#issuecomment-977597528)

*/

var leadingLine = "\r\n";    // either; "" = no leading line, or "\r\n" = leading line

foreach (var m in Model.AllMeasures) { m.Expression = leadingLine + FormatDax(m.Expression); }

foreach (var i in Model.AllCalculationItems) { i.Expression = leadingLine + FormatDax(i.Expression); }

foreach (var t in Model.Tables.Select(x => x as CalculatedTable).Where(x => x != null)) { t.Expression = leadingLine + FormatDax(t.Expression); }

foreach (var c in Model.AllColumns.Select(x => x as CalculatedColumn).Where(x => x != null)) { c.Expression = leadingLine + FormatDax(c.Expression); }

Info("Script finished.");
```

**Comment by samaguire** [comment]
> There is also a bug after a /* comment */ line.
> 
> I kept pressing the formatter and it kept adding new rows and never removing them. The only fix is to manually remove them, or switch back to DAXFormatter.com which removed all of it.

Just confirming I get the same issue with multiline comment blocks: adding two new lines at the end of the block and indenting the body and close of the block each time the formatter is applied.

**original**

![image](https://user-images.githubusercontent.com/62320770/144322849-d927949e-f270-4b4d-b9ef-1a6369c0bf59.png)

**first run**

![image](https://user-images.githubusercontent.com/62320770/144322869-afbb05b3-0c87-4a71-8cc2-021a3f755094.png)

**second run**

![image](https://user-images.githubusercontent.com/62320770/144322898-7846dd08-659b-4b25-a1a6-7f1e1f334128.png)

**third run**

![image](https://user-images.githubusercontent.com/62320770/144323344-9f97da5f-5bd4-4fa5-963f-4e029c66ba70.png)

**Comment by otykier** [comment]
Fixed in [3.1.7](https://github.com/TabularEditor/TabularEditor3/releases/tag/3.1.7). Thanks for reporting!
